cmake: Check user-defined `APPEND_*FLAGS` variables early

hebasto opened the pull request:
For the purpose of checks performed by the build system, we strive to handle user-defined `APPEND_*FLAGS` in the same way as flags provided by other standard means. In particular, they are considered when the `try_compile()` command is used.

However, these flags are not considered during `enable_language()` command invocation due to design limitations, which might cause issues, such as https://github.com/bitcoin/bitcoin/issues/32323.

This PR addresses the issue by introducing an additional compiler check that does consider the user-defined `APPEND_*FLAGS`.

Fixes https://github.com/bitcoin/bitcoin/issues/32323:
```
$ cmake -B build -DAPPEND_CXXFLAGS="-ftrivial-auto-var-init"
-- The CXX compiler identification is GNU 14.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CXX_COMPILER_WORKS
-- Performing Test CXX_COMPILER_WORKS - Failed
CMake Error at cmake/module/EnableLanguage.cmake:42 (message):
  The CXX compiler is not able to compile a simple test program.

  Check that the "APPEND_*FLAGS" variables are set correctly.



Call Stack (most recent call first):
  CMakeLists.txt:71 (bitcoincore_enable_language)


-- Configuring incomplete, errors occurred!
hebasto@TS-P340:~/dev/bitcoin$ cmake -B build -DAPPEND_CXXFLAGS="-ftrivial-auto-var-init=zero"
-- Performing Test CXX_COMPILER_WORKS
-- Performing Test CXX_COMPILER_WORKS - Success
...
```

janb84 reviewed (commented):
ACK [eb2a943](https://github.com/bitcoin/bitcoin/commit/eb2a9435044ac91442fafc606c5af2f473bff3c5)

- PR fixes #32323 (reproduced on main, checked it was solved by this PR) ✅
- Code review looks good to me. Adds a new macro with additional checks and some optimisations.✅ (There is one reference to `enable_languages()` in secp256k1 but that is separate and not relevant to this PR )

<details>

### Main:

Configure with faulty flag: 
`cmake -B build -DAPPEND_CXXFLAGS="-ftrivial-auto-var-init"`

expected config BREAK

Configure with correct flag: 
`cmake -B build -DAPPEND_CXXFLAGS="-ftrivial-auto-var-init=zero"`
not expected: STILL BREAKS

### This PR:

Configure with faulty flag 
`cmake -B build -DAPPEND_CXXFLAGS="-ftrivial-auto-var-init"`
expected BREAK

Configure with correct flag 
`cmake -B build -DAPPEND_CXXFLAGS="-ftrivial-auto-var-init=zero"`
config WORKS

</details>

BrandonOdiwuor reviewed (commented):
ACK eb2a9435044ac91442fafc606c5af2f473bff3c5

Confirmed that the new `bitcoincore_enable_language(...)` macro properly validates user-defined `APPEND_*FLAGS` early via `check_source_compiles(...)`, fixing #32323 by catching bad flags during enable_language(...), and removing the need a clean build directory.

(system: macOS 15.6)

(branch: master)
<img width="1512" height="917" alt="Screenshot 2025-08-21 at 13 21 10" src="https://github.com/user-attachments/assets/be10daae-31c4-4eb7-8504-51fb0a0d72ec" />

(commit: eb2a9435044ac91442fafc606c5af2f473bff3c5)
<img width="1512" height="914" alt="Screenshot 2025-08-21 at 13 19 51" src="https://github.com/user-attachments/assets/59de425d-6a48-4ce1-86c2-8378feedbd77" />

purpleKarrot commented:
The release notes for 29.0 contain:

> 7. If any of the `CPPFLAGS`, `CFLAGS`, `CXXFLAGS` or `LDFLAGS` environment variables were used in your Autotools-based build process, you should instead use the corresponding CMake variables (`APPEND_CPPFLAGS`, `APPEND_CFLAGS`, `APPEND_CXXFLAGS` and `APPEND_LDFLAGS`).

CMake does not use the `APPEND_*FLAGS` by default, but it actually does use the mentioned environment variables. But why is it mentioned that they should not be used?

purpleKarrot commented:
> The expected behavior is as follows:
> 
> ```
> $ cmake -B build -DAPPEND_CXXFLAGS="-mno-sse4.1"
> <snip>
> -- Performing Test HAVE_SSE41
> -- Performing Test HAVE_SSE41 - Failed
> <snip>
> ```

I see, you want `CXXFLAGS` to affect the SIMD capability detection. But is that the right approach?

We can search for the existence of Qt and based on the result we can enable building of the GUI. Or, we can provide an option whether the GUI should be built and based on that setting, we make the existence of Qt a requirement.
But we would not search for Qt headers and libraries in the paths that the user passes in `-I` and `-L` flags of `CXXFLAGS` and `LDFLAGS`, right?

In the same sense, we should check for SIMD capabilities and based on the result provide options to use them. Or, provide the options unconditionally and based on the setting make the capability a requirement. But relying on `CXXFLAGS` or `APPEND_CXXFLAGS` does not seem the right approach to me.

Also, it requires custom procedural logic like `CheckSourceCompilesWithFlags` that would not be necessary if we used a more standard approach.

hebasto committed:
cmake: Wrap `enable_language()` command in macro

This change introduces the `bitcoincore_enable_language()` macro to wrap
CMake’s `enable_language()` command, applying Bitcoin Core-specific
settings and checks.

hebasto committed:
cmake: Move `APPEND_*` variables logic to `bitcoincore_enable_language`

This change is required for the following commit.

hebasto committed:
cmake: Check user-defined `APPEND_*` variables early

This change improves usability in cases where the user provides
`APPEND_*` variables that are incompatible with the compiler for some
reason.

hebasto commented:
Rebased to resolve conflicts.

hebasto commented:
> But relying on `CXXFLAGS` or `APPEND_CXXFLAGS` does not seem the right approach to me.

This PR is specifically focused on resolving the current [issue](https://github.com/bitcoin/bitcoin/issues/32323). I believe that the broader discussion regarding `APPEND_*FLAGS` should be continued elsewhere.

purpleKarrot commented on cmake/module/EnableLanguage.cmake:
```diff
@@ -0,0 +1,25 @@
+      # So disable this tool check in further enable_language() commands.
+      set(CMAKE_INSTALL_NAME_TOOL "${CMAKE_COMMAND} -E true")
+    endif()
+    enable_language(${language})
```
CMake versions earlier than 4.5 (which is not released at the time of writing this comment) require that the `enable_languages()` command is called at file scope, "in the highest directory common to all targets using the named language directly for compiling sources or indirectly through link dependencies".

The simplest approach is not to use this command at all, and instead enable all required languages in the top-level `project()` command.

purpleKarrot commented:
> This PR is specifically focused on resolving the current https://github.com/bitcoin/bitcoin/issues/32323. I believe that the broader discussion regarding APPEND_*FLAGS should be continued elsewhere.

The broader discussion whether `APPEND_*FLAGS` is a feature worth keeping should continue elsewhere.
But the question whether `APPEND_*FLAGS` (or `CFLAGS`, `CXXFLAGS`, for that matter) should influence any kind of capability detection is very relevant to the question whether the proposed code is the best way to solve that issue.

My position is "no". And my rationale is that compiler capability detection is cached. We don't want to add more ways for the cache to go stale.

fanquake commented:
So what was the resolution here?

maflcko commented:
> So what was the resolution here?

I left a comment in the issue: `rm -rf ...` should work fine here and in other cases?
